pkg/ccl/logictestccl/tests/local-repeatable-read/local-repeatable-read_test: TestRepeatableReadLogic_schema_repair failed

pkg/ccl/logictestccl/tests/local-repeatable-read/local-repeatable-read_test.TestRepeatableReadLogic_schema_repair [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20899178?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20899178?buildTab=artifacts#/) on release-26.1 @ [6eafdbb7f69ff1f880d992b1af3333ab4617893f](https://github.com/cockroachdb/cockroach/commits/6eafdbb7f69ff1f880d992b1af3333ab4617893f):


```
google.golang.org/grpc/internal/transport.(*controlBuffer).get(0x4006b08460, 0x1)
	external/org_golang_google_grpc/internal/transport/controlbuf.go:418 +0x140
google.golang.org/grpc/internal/transport.(*loopyWriter).run(0x4001577180)
	external/org_golang_google_grpc/internal/transport/controlbuf.go:552 +0x6c
google.golang.org/grpc/internal/transport.NewServerTransport.func2()
	external/org_golang_google_grpc/internal/transport/http2_server.go:336 +0xcc
created by google.golang.org/grpc/internal/transport.NewServerTransport in goroutine 22358
	external/org_golang_google_grpc/internal/transport/http2_server.go:333 +0x13c0

goroutine 21888 [sync.WaitGroup.Wait, 14 minutes, "job":"AUTO SPAN CONFIG RECONCILIATION id=1134994007375970305", "n":"1", "rangefeed":"sql-watcher-zones-rangefeed"]:
sync.runtime_SemacquireWaitGroup(0x40069a9a18?, 0xbc?)
	GOROOT/src/runtime/sema.go:114 +0x38
sync.(*WaitGroup).Wait(0x4004447d88)
	GOROOT/src/sync/waitgroup.go:206 +0xa8
golang.org/x/sync/errgroup.(*Group).Wait(0x4004447d80)
	external/org_golang_x_sync/errgroup/errgroup.go:56 +0x28
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0x4004447d80?, {0x7b38ef8?, 0x4003297090?}, 0x4007a31f38?})
	pkg/util/ctxgroup/ctxgroup.go:148 +0x64
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0x7b38f30?, 0x400202a3c0?}, {0x40069a9c50, 0x2, 0x5f07400?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x100
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run(0x40078de000, {0x7b38f30, 0x400202a3c0}, {0x7ba1b40, 0x4003296eb0}, 0x0)
	pkg/kv/kvclient/rangefeed/rangefeed.go:396 +0x8e8
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).start.func1({0x7b38ef8?, 0x4003296f00?})
	pkg/kv/kvclient/rangefeed/rangefeed.go:256 +0x270
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x7b38ef8, 0x4003296f00})
	pkg/util/stop/stopper.go:440 +0x7c
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 21782
	pkg/util/stop/stopper.go:438 +0xe0

goroutine 22352 [select, 14 minutes, "tags":"n1,rnode=1,raddr=127.0.0.1:39103,class=system,rpc", "workload.id":"RPC_HEARTBEAT"]:
google.golang.org/grpc/internal/grpcsync.(*CallbackSerializer).run(0x400220e1c0, {0x7b38ef8, 0x400b53dd60})
	external/org_golang_google_grpc/internal/grpcsync/callback_serializer.go:83 +0xbc
created by google.golang.org/grpc/internal/grpcsync.NewCallbackSerializer in goroutine 22349
	external/org_golang_google_grpc/internal/grpcsync/callback_serializer.go:55 +0x10c

goroutine 27209 [chan receive, 14 minutes, "job":"MVCC STATISTICS UPDATE id=104", "n":"1"]:
github.com/cockroachdb/cockroach/pkg/sql.(*mvccStatisticsUpdateJob).Resume(0x4007798bd0, {0x7b38f30, 0x400258d7a0}, {0x66f9e80, 0x400b70e250})
	pkg/sql/mvcc_statistics_update_job.go:85 +0x130
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0x4001d73e00?, 0x0?, 0x4009acd8d0, {0x7b1f100?, 0x4007798bd0?}, {0x7b38f30?, 0x400258d7a0?}, {0x66f9e80?, 0x400b70e250?})
	pkg/jobs/registry.go:1647 +0xe8
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0x4009509008, {0x7b38ef8, 0x40073022d0}, {0x66f9e80, 0x400b70e250}, {0x7b1f100, 0x4007798bd0}, 0x4001d73e00, {0x40061243c2, 0x7}, ...)
	pkg/jobs/registry.go:1648 +0x870
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0x4009509008, {0x7b38ef8, 0x40073022d0}, {0x7b1f100, 0x4007798bd0}, 0x4001d73e00, {0x40061243c2, 0x7}, {0x40055722a0, 0x7})
	pkg/jobs/adopt.go:418 +0x35c
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x7b38f30?, 0x400258d170?})
	pkg/jobs/adopt.go:271 +0xe0
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x7b38f30, 0x400258d170})
	pkg/util/stop/stopper.go:440 +0x7c
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 27118
	pkg/util/stop/stopper.go:438 +0xe0
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRepeatableReadLogic_schema_repair.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58122

```
panic: test timed out after 15m0s
› running tests:
› › TestRepeatableReadLogic_schema_repair (14m45s)
› › TestRepeatableReadLogic_schema_repair/force_delete_data (14m43s)
```
looks like we're not stuck but the lease code might be slow (?)
```
goroutine 22502 [runnable]:
regexp.(*Regexp).tryBacktrack(0x40011c6780?, 0x400631a580?, {0x7b63820?, 0x400631a600?}, 0x1?, 0x4?)
› GOROOT/src/regexp/backtrack.go:138 +0xa74
regexp.(*Regexp).backtrack(0x40011c6780, {0x0, 0x0, 0x0}, {0x40020bbf00, 0x11}, 0x0, 0x2, {0x400ada05a0, 0x0, ...})
› GOROOT/src/regexp/backtrack.go:351 +0x2a4
regexp.(*Regexp).doExecute(0x0?, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x40020bbf00, 0x11}, 0x0, 0x2, ...)
› GOROOT/src/regexp/exec.go:535 +0x1c8
regexp.(*Regexp).replaceAll(0x40011c6780, {0x0, 0x0, 0x0}, {0x40020bbf00, 0x11}, 0x2, 0x4009128398)
› GOROOT/src/regexp/regexp.go:619 +0x110
regexp.(*Regexp).ReplaceAllString(0x40011c6780, {0x40020bbf00, 0x11}, {0x0, 0x0})
› GOROOT/src/regexp/regexp.go:577 +0xb0
github.com/cockroachdb/redact/internal/markers.RedactableString.StripMarkers(...)
› external/com_github_cockroachdb_redact/internal/markers/markers.go:33
github.com/cockroachdb/errors/errutil.(*leafError).Error(0x4009128488?)
› external/com_github_cockroachdb_errors/errutil/redactable.go:24 +0x30
github.com/cockroachdb/errors/withstack.(*withStack).Error(0x4009128498?)
› external/com_github_cockroachdb_errors/withstack/withstack.go:64 +0x24
github.com/cockroachdb/errors/markers.safeGetErrMsg({0x7ad2da0?, 0x400216e5a0?})
› external/com_github_cockroachdb_errors/markers/markers.go:217 +0x64
github.com/cockroachdb/errors/markers.isMarkEqual({0x7ad2da0, 0x400216e5d0}, {0x7ad2da0, 0x400216e5a0})
› external/com_github_cockroachdb_errors/markers/markers.go:104 +0x2ac
github.com/cockroachdb/errors/markers.Is({0x7ad2da0, 0x400216e5d0}, {0x7ad2da0, 0x400216e5a0})
› external/com_github_cockroachdb_errors/markers/markers.go:72 +0x210
github.com/cockroachdb/errors.Is(...)
› external/com_github_cockroachdb_errors/markers_api.go:35
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).Acquire(0x4001282848, {0x7b38ef8, 0x40054185f0}, {0x7b12190, 0x4001e6a390}, 0x6c)
› pkg/sql/catalog/lease/lease.go:2203 +0xac
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).AcquireByName(0x4001282848, {0x7b38ef8, 0x40054185f0}, {0x7b12190, 0x4001e6a390}, 0x68, 0x69, {0x400705bc7d, 0x11})
› pkg/sql/catalog/lease/lease.go:2057 +0x2d0
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*leasedDescriptors).getByName(0x4006d8d3a8, {0x7b38ef8, 0x40054185f0}, {0x7b20ae0, 0x4003485760}, 0x68, 0x69, {0x400705bc7d, 0x11}, 0x0)
› pkg/sql/catalog/descs/leased_descriptors.go:281 +0x1e0
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getNonVirtualDescriptorID.func6()
› pkg/sql/catalog/descs/descriptor.go:656 +0x154
...
```

--- a/pkg/sql/catalog/lease/descriptor_state.go
+++ b/pkg/sql/catalog/lease/descriptor_state.go
@@ -98,8 +98,10 @@ func (t *descriptorState) findForTimestamp(
 	if len(t.mu.active.data) == 0 {
 		// If the descriptor is marked as offline, we should attempt
 		// a historical query to fetch it, since it's likely querying
-		// the past. A non-historical query will fail instantly.
-		if t.mu.takenOffline {
+		// the past. If the requested time is past the offline time,
+		// then attempt to renew again, since it could have come online
+		// again.
+		if t.mu.takenOffline && timestamp.GetTimestamp().Less(t.mu.takenOfflineAt) {
 			return nil, false, errReadOlderVersion
 		}
 		return nil, false, errRenewLease
@@ -130,7 +132,6 @@ func (t *descriptorState) findForTimestampImpl(
 			if err := catalog.FilterDroppedDescriptor(desc); err != nil {
 				return nil, false, err
 			}
-
 			latest := i+1 == len(t.mu.active.data)
 			if !desc.hasExpired(ctx, readTimestamp) {
 				// Existing valid descriptor version.

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -948,26 +948,41 @@ func (m *Manager) readOlderVersionForTimestamp(
 		// We permit gaps in historical versions. We want to find the timestamp
 		// that represents the start of the validity interval for the known version
 		// which immediately follows the timestamps we're searching for.
-		i := sort.Search(len(t.mu.active.data), func(i int) bool {
+		tsForComparison := timestamp.GetTimestamp()
+		indexAfterTS := sort.Search(len(t.mu.active.data), func(i int) bool {
 			return timestamp.GetTimestamp().Less(t.mu.active.data[i].GetModificationTime())
 		})
+
+		// If the read timestamp and base timestamp are different, additionally,
+		// check if we need a slightly later end timestamp. This guarantees we will
+		// populate every possible descriptor needed for this request.
+		if timestamp.GetTimestamp() != timestamp.GetBaseTimestamp() {
+			baseTimeStampIndex := sort.Search(len(t.mu.active.data), func(i int) bool {
+				return timestamp.GetBaseTimestamp().Less(t.mu.active.data[i].GetModificationTime())
+			})
+			if baseTimeStampIndex != len(t.mu.active.data) {
+				indexAfterTS = baseTimeStampIndex
+				tsForComparison = timestamp.GetBaseTimestamp()
+			}
+		}
 
 		// If the timestamp we're searching for is somehow after the last descriptor
 		// we have in play, then either we have the right descriptor, or some other
 		// shenanigans where we've evicted the descriptor has occurred.
-		//
-		// TODO(ajwerner): When we come to modify this code to allow us to find
-		// historical descriptors which have been dropped, we'll need to rework
-		// this case and support providing no upperBound to
-		// getDescriptorFromStoreForInterval.
-		if i == len(t.mu.active.data) ||
+		if indexAfterTS == len(t.mu.active.data) ||
 			// If we found a descriptor that isn't the first descriptor, go and check
 			// whether the descriptor for which we're searching actually exists. This
 			// will deal with cases where a concurrent fetch filled it in for us.
-			i > 0 && timestamp.GetTimestamp().Less(t.mu.active.data[i-1].getExpiration(ctx)) {
-			return hlc.Timestamp{}, false, true
+			indexAfterTS > 0 && tsForComparison.Less(t.mu.active.data[indexAfterTS-1].getExpiration(ctx)) {
+			// If the descriptor is offline, then nothing newer can exist.
+			if !t.mu.takenOffline {
+				return hlc.Timestamp{}, false, true
+			}
+			// Otherwise, the table is offline, so there may be newer descriptor
+			// versions before the offline timestamp.
+			return t.mu.takenOfflineAt, true, false
 		}
-		return t.mu.active.data[i].GetModificationTime(), false, false
+		return t.mu.active.data[indexAfterTS].GetModificationTime(), false, false
 	}()
 	if done {
 		return nil, nil

